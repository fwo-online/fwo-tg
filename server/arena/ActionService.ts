import arena from '@/arena';
import type { Action } from '@fwo/shared';

export type ActionKey = keyof typeof arena.actions;

export class ActionService {
  static isAction(maybeAction: string): maybeAction is ActionKey {
    return maybeAction in arena.actions;
  }

  static isBaseAction(action: string) {
    if (ActionService.isAction(action)) {
      return arena.actions[action].lvl === 0;
    }

    return false;
  }

  static toObject(action: ActionKey, lvl = 0): Action {
    const c = arena.actions[action];

    if ('cost' in c && 'costType' in c) {
      const cost = Array.isArray(c.cost) ? c.cost[lvl - 1] : c.cost;

      return {
        name: c.name,
        displayName: c.displayName,
        costType: c.costType,
        cost,
        orderType: c.orderType,
      };
    }

    return {
      name: c.name,
      displayName: c.displayName,
      orderType: c.orderType,
    };
  }
}
