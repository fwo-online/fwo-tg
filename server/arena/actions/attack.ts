import type { Affect } from '../Constuructors/interfaces/Affect';
import PhysConstructor from '../Constuructors/PhysConstructor';
import CastError from '../errors/CastError';
/**
 * Физическая атака
 */
class Attack extends PhysConstructor implements Affect {
  constructor() {
    super({
      name: 'attack',
      displayName: 'Атака',
      desc: 'Атака',
      lvl: 0,
      orderType: 'all',
    });
  }

  /**
 * Кастомный обработчик атаки
 */
  run() {
    if (!this.params) {
      return;
    }
    const { initiator, target } = this.params;

    target.flags.isHited = {
      initiator: initiator.nick, val: this.status.effect,
    };

    target.stats.down('hp', this.status.effect);
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isHited) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}
export default new Attack();
