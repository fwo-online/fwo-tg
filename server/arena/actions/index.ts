export { default as attack } from './attack';
export { default as handsHeal } from './handsHeal';
export { default as protect } from './protect';
export { default as regeneration } from './regeneration';
