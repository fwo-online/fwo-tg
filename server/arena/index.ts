import type { Clan } from '@/models/clan';
import type { ItemOutput } from '@fwo/shared';
import type { ItemSet } from '@fwo/shared';
import type { CharacterService } from './CharacterService';
import type GameService from './GameService';
import type MatchMaking from './MatchMakingService';
import type * as actions from './actions';
import type * as magics from './magics';
import type * as passiveSkills from './passiveSkills';
import type * as skills from './skills';
import type * as weaponMastery from './weaponMastery';

export default {
  mm: null as unknown as typeof MatchMaking,
  characters: {} as Record<string, CharacterService>,
  games: {} as Partial<Record<string, GameService>>,
  items: {} as unknown as Record<string, ItemOutput>,
  itemsSets: {} as unknown as Record<string, ItemSet>,
  magics: {} as unknown as typeof magics,
  skills: {} as unknown as typeof skills,
  actions: {} as unknown as typeof actions &
    typeof magics &
    typeof skills &
    typeof passiveSkills &
    typeof weaponMastery,
  clans: new Map<string, Clan>(),
};
