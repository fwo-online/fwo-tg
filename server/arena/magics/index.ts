/**
 * Первичный загрузчик магий
 */
export { default as acidSpittle } from './acidSpittle';
export { default as anathema } from './anathema';
export { default as blessing } from './blessing';
export { default as blight } from './blight';
export { default as bodySpirit } from './bodySpirit';
export { default as chainLightning } from './chainLightning';
export { default as curse } from './curse';
export { default as dispel } from './dispel';
export { default as dustShield } from './dustShield';
export { default as eclipse } from './eclipse';
export { default as entangle } from './entangle';
export { default as exorcism } from './exorcism';
export { default as fireBall } from './fireBall';
export { default as frostTouch } from './frostTouch';
export { default as fireRain } from './fireRain';
export { default as glitch } from './glitch';
export { lightHeal } from './lightHeal';
export { default as lightShield } from './lightShield';
export { default as madness } from './madness';
export { default as magicArmor } from './magicArmor';
export { default as magicArrow } from './magicArrow';
export { default as magicDefense } from './magicDefense';
export { default as mediumAura } from './mediumAura';
export { default as mediumHeal } from './mediumHeal';
export { default as magicWall } from './magicWall';
export { default as paralysis } from './paralysis';
export { default as poisonBreath } from './poisonBreath';
export { default as physicalSadness } from './physicalSadness';
export { default as rockfall } from './rockfall';
export { default as silence } from './silence';
export { default as smallAura } from './smallAura';
export { default as stoneSkin } from './stoneSkin';
export { default as strongHeal } from './strongHeal';
export { default as secondLife } from './secondLife';
export { default as sleep } from './sleep';
export { default as strongAura } from './strongAura';
export { default as vampirism } from './vampirism';
