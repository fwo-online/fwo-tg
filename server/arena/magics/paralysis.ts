import { CommonMagic } from '../Constuructors/CommonMagicConstructor';
import type { Affect } from '../Constuructors/interfaces/Affect';
import CastError from '../errors/CastError';

/**
 * Паралич
 * Основное описание магии общее требовани есть в конструкторе
 */
class Paralysis extends CommonMagic implements Affect {
  constructor() {
    super({
      name: 'paralysis',
      displayName: 'Паралич',
      desc: 'Парализует цель',
      cost: 8,
      baseExp: 80,
      costType: 'mp',
      lvl: 2,
      orderType: 'enemy',
      aoeType: 'target',
      magType: 'bad',
      chance: ['1d80', '1d90', '1d100'],
      profList: ['p'],
      effect: [],
    });
  }

  run() {
    const { target } = this.params;
    target.flags.isParalysed = true;
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isParalysed) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}

export default new Paralysis();
