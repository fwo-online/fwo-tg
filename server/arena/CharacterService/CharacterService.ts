import _ from 'lodash';
import type { UpdateQuery } from 'mongoose';
import { findCharacter, removeCharacter, updateCharacter } from '@/api/character';
import arena from '@/arena';
import config from '@/arena/config';
import { InventoryService } from '@/arena/InventoryService';
import type { HarksLvl } from '@/data/harks';
import type { Char } from '@/models/character';
import type { Character, CharacterClass, CharacterPublic } from '@fwo/shared';
import { assignWithSum } from '@/utils/assignWithSum';
import { calculateDynamicAttributes } from './utils/calculate-dynamic-attributes';
import { sum } from 'es-toolkit';
import ValidationError from '@/arena/errors/ValidationError';

/**
 * Конструктор персонажа
 * @todo сюда нужны будет get/set функции для intreface части
 * @todo Сейчас массив arena.player не является массивом объектов Character,
 * нужно переработать
 */

/**
 * Класс описывающий персонажа внутри игры
 */
export class CharacterService {
  tempHarks: HarksLvl & { free: number };
  mm: { status?: string; time?: number };
  inventory: InventoryService;

  /**
   * Конструктор игрока
   */
  constructor(public charObj: Char) {
    this.inventory = new InventoryService(charObj, charObj.inventory ?? []);
    this.charObj = charObj;
    this.tempHarks = {
      ...charObj.harks,
      free: charObj.free,
    };
    this.mm = {};
    this.resetExpLimit();
  }

  wasLvlUp = false;
  autoreg = false;

  get id() {
    return this.charObj.id;
  }

  get prof() {
    return this.charObj.prof;
  }

  get class() {
    return this.charObj.prof as CharacterClass;
  }

  get lvl() {
    return this.charObj.lvl;
  }

  get owner() {
    return this.charObj.owner;
  }

  get nickname() {
    return this.charObj.nickname;
  }

  get gold() {
    return this.charObj.gold;
  }

  set gold(value) {
    this.charObj.gold = value;
  }

  get exp() {
    return this.charObj.exp;
  }

  set exp(value) {
    this.resetExpLimit();
    this.bonus += Math.round(value / 100) - Math.round(this.charObj.exp / 100);
    this.charObj.exp = value;
    void this.addLvl();
  }

  set expEarnedToday(value) {
    this.charObj.expLimit.earn = value;
  }

  get expEarnedToday() {
    return this.charObj.expLimit.earn;
  }

  get expLimitToday() {
    return this.lvl * config.lvlRatio * 2000;
  }

  get games() {
    return this.charObj.statistics.games;
  }

  get kills() {
    return this.charObj.statistics.kills;
  }

  get runs() {
    return this.charObj.statistics.runs;
  }

  get death() {
    return this.charObj.statistics.death;
  }

  get free() {
    return this.tempHarks.free;
  }

  set free(value) {
    this.charObj.free = value;
    this.tempHarks.free = value;
  }

  // Базовые harks без учёта надетых вещей
  get attributes() {
    return this.charObj.harks;
  }

  get magics() {
    return this.charObj.magics || {};
  }

  get skills() {
    return this.charObj.skills || {};
  }

  get bonus() {
    return this.charObj.bonus;
  }

  set bonus(value) {
    this.charObj.bonus = value;
  }

  get clan() {
    return this.charObj.clan;
  }

  /** Суммарное количество опыта, требуемое для следующего уровня */
  get nextLvlExp() {
    return 2 ** (this.lvl - 1) * 1000 * config.lvlRatio;
  }

  get favoriteMagicList() {
    return this.charObj.favoriteMagicList;
  }

  /** @type {string[]} */
  set favoriteMagicList(value) {
    this.charObj.favoriteMagicList = value;
  }

  // Суммарный объект характеристик + вещей.
  getDynamicAttributes(attributes = this.attributes) {
    const characterAttributes = structuredClone(attributes);
    const inventoryAttributes = this.inventory.attributes;

    assignWithSum(characterAttributes, inventoryAttributes);
    const dynamicHarks = calculateDynamicAttributes(this, characterAttributes);
    assignWithSum(dynamicHarks, this.inventory.harksFromItems);
    return dynamicHarks;
  }

  resetExpLimit() {
    const date = new Date();
    if (date > this.charObj.expLimit.expiresAt) {
      date.setUTCHours(24, 0, 0, 0);
      this.charObj.expLimit.expiresAt = date;
      this.charObj.expLimit.earn = 0;
    }
  }

  /**
   * @param {Partial<Statistics>} stat
   */
  addGameStat(stat) {
    _.forEach(stat, (val, key) => {
      this.charObj.statistics[key] += val;
    });
  }

  /**
   * @param {string} reason
   */
  getPenaltyDate(reason) {
    const penalty = this.charObj.penalty.find((p) => p.reason === reason);
    if (penalty && penalty.date.valueOf() > Date.now()) {
      return penalty.date;
    }
    return false;
  }

  /**
   * @param {string} reason
   * @param {number} minutes
   */
  async updatePenalty(reason, minutes) {
    const date = new Date();
    date.setHours(date.getHours(), date.getMinutes() + minutes);

    const penalty = { reason, date };
    const index = this.charObj.penalty.findIndex((p) => p.reason === reason);

    if (index === -1) {
      this.charObj.penalty.push(penalty);
    } else {
      this.charObj.penalty[index] = penalty;
    }
    await this.saveToDb();
  }

  /**
   * Проверяет количество опыта для следующего уровня. Добавляет уровень, если опыта достаточно
   * @returns {void}
   */
  async addLvl() {
    if (this.exp >= this.nextLvlExp) {
      this.charObj.lvl += 1;
      this.free += 10;
      await this.addLvl();
      this.wasLvlUp = true;
    } else {
      await this.saveToDb();
    }
  }

  // В функциях прокачки харок следует использоваться this.charObj.harks
  getIncreaseHarkCount(hark) {
    const count = this.tempHarks[hark] - this.charObj.harks[hark];
    return count || '';
  }

  increaseHark(harkName) {
    if (this.tempHarks.free < 1) {
      throw Error('Недостаточно очков');
    }

    this.tempHarks[harkName] += 1;
    this.tempHarks.free -= 1;
  }

  resetHarks() {
    this.tempHarks = {
      ...this.charObj.harks,
      free: this.charObj.free,
    };
  }

  async increaseHarks(harks: HarksLvl) {
    const isValid = Object.entries(harks).every(([key, value]) => value >= this.charObj.harks[key]);
    if (!isValid) {
      throw new ValidationError('Аттрибут не может быть уменьшен');
    }

    const free = this.free - (sum(Object.values(harks)) - sum(Object.values(this.charObj.harks)));
    if (free < 0) {
      throw new ValidationError('Недостаточно очков');
    }

    this.charObj.harks = harks;
    this.charObj.free = free;

    await this.save({ harks, free });
  }

  async submitIncreaseHarks({ free, ...harks } = this.tempHarks) {
    this.charObj.harks = harks;
    this.charObj.free = free;

    // @todo сюда нужно будет предусмотреть проверки на корректность сохраняемых данных
    return this.save({ harks, free });
  }

  async buyItem(itemCode: string) {
    const item = arena.items[itemCode];

    if (this.gold < item.price) {
      throw new ValidationError('Недостаточно золота');
    }

    this.gold -= item.price;
    await this.inventory.addItem(itemCode);
    return this.saveToDb();
  }

  /**
   * Продажа предмета.
   */
  async sellItem(inventoryID: string) {
    const inventory = await this.inventory.removeItem(inventoryID);
    const item = arena.items[inventory.code];

    this.gold += item.price / 2;
    await this.saveToDb();
  }

  /**
   * @param {Clan} clan
   */
  async joinClan(clan) {
    this.charObj.clan = clan;
    await this.saveToDb();
    const char = await CharacterService.getCharacter(this.owner);
    return char;
  }

  async leaveClan() {
    this.charObj.clan = undefined;
    await this.updatePenalty('clan_leave', 5 * 24 * 60);
  }

  /**
   * @desc Получает идентификатор игры из charId участника
   * @return gameId идентификатор игры
   */
  get gameId() {
    return this.mm.status || '';
  }

  /**
   * Setter gameId
   * @param newStatus новый статус mm персонажа
   */
  set gameId(newStatus) {
    this.mm.status = newStatus;
  }

  get currentGame() {
    return arena.games[this.gameId];
  }

  /**
   * Загрузка чара в память
   * @param {string} owner идентификатор пользователя в TG (tgId)
   */
  static async getCharacter(owner: string) {
    const charFromDb = await findCharacter({ owner });

    const char = new CharacterService(charFromDb);
    arena.characters[char.id] = char;
    return char;
  }

  /**
   * Загрузка чара в память
   * @param {string} id идентификатор пользователя
   */
  static async getCharacterById(id: string) {
    const cachedChar = arena.characters[id];
    if (cachedChar) {
      return cachedChar;
    }

    const charFromDb = await findCharacter({ _id: id });

    const char = new CharacterService(charFromDb);
    arena.characters[id] = char;
    return char;
  }

  /**
   * Возвращает объект игры по Id чара
   * @param {String} charId идентификатор чара;
   * @return {this} Объект игры
   * @todo нужно перенести это не в static а во внутрь экземпляра класса
   */
  static getGameFromCharId(charId) {
    const { gameId } = arena.characters[charId];
    if (arena.games[gameId]) {
      return arena.games[gameId];
    }
    throw Error('gameId_error');
  }

  /**
   * Функция получения новой магии
   * @param magicId идентификатор магии
   * @param lvl уровень проученной магии
   */
  async learnMagic(magicId: string, lvl: number) {
    this.magics[magicId] = lvl;
    // опасный тест
    await this.saveToDb();
  }

  /**
   * Получение нового умения
   * @param {string} skillId идентификатор умения
   * @param {number} lvl уровень проученного умения
   */
  async learnSkill(skillId, lvl) {
    this.skills[skillId] = lvl;
    await this.saveToDb();
  }

  /**
   * Сохраняет состояние чара в базу
   * @param {import('mongoose').UpdateQuery<import ('@/models/character').Char>} [query]
   */
  async save(query: UpdateQuery<Char>) {
    console.log('Saving char :: id', this.id);
    try {
      return await updateCharacter(this.id, query);
    } catch (e) {
      console.error('Fail on CharSave:', e);
    }
  }

  /**
   * Сохраняет состояние чара в базу
   * @todo важная функция которая сохраняет параметры чара в базу
   * @deprecated
   */
  async saveToDb() {
    try {
      console.log('Saving char :: id', this.id);
      const { gold, exp, magics, bonus, skills, lvl, clan, free } = this;
      return await updateCharacter(this.id, {
        gold,
        exp,
        magics,
        bonus,
        skills,
        lvl,
        clan,
        penalty: this.charObj.penalty,
        free,
        expLimit: this.charObj.expLimit,
        statistics: this.charObj.statistics,
        favoriteMagicList: this.charObj.favoriteMagicList,
        inventory: this.inventory.inventory,
      });
    } catch (e) {
      console.error('Fail on CharSave:', e);
    }
  }

  async remove() {
    await removeCharacter(this.owner);
    delete arena.characters[this.id];
  }

  toObject(): Character {
    return {
      id: this.id,
      owner: this.owner,
      name: this.nickname,
      class: this.prof as CharacterClass,
      attributes: this.attributes,
      magics: this.magics,
      skills: this.skills,
      clan: this.clan,
      inventory: this.inventory.toObject(),
      free: this.charObj.free,
      bonus: this.bonus,
      gold: this.gold,
      lvl: this.lvl,
      exp: this.exp,
      dynamicAttributes: this.getDynamicAttributes(),
      game: this.currentGame?.info.id,
    };
  }

  toPublicObject(): CharacterPublic {
    return {
      id: this.id,
      name: this.nickname,
      class: this.prof as CharacterClass,
      lvl: this.lvl,
      clan: this.clan,
    };
  }
}
