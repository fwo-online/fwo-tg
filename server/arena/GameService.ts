import { createGame } from '@/api/game';
import type { Game } from '@/models/game';
import type { LongItem } from '@/arena/Constuructors/LongMagicConstructor';
import { engine } from '@/arena/EngineService';
import { HistoryService, type HistoryItem } from '@/arena/HistoryService';
import type * as magics from '@/arena/magics';
import OrderService from '@/arena/OrderService';
import PlayersService, { type Player } from '@/arena/PlayersService';
import { RoundService, RoundStatus } from '@/arena/RoundService';
import arena from '@/arena';
import { mapValues } from 'es-toolkit';
import EventEmitter from 'node:events';
import { type GameStatus, reservedClanName } from '@fwo/shared';
import { getRandomComponent } from '@/utils/getRandomComponent';

export type KickReason = 'afk' | 'run';

export interface GlobalFlags {
  isEclipsed: { initiator: Player }[];
}

/**
 * GameService
 *
 * @description Обработка около игровой логики
 * @module Service/Game
 * @todo сейчас после того как Player отключился, socket выходит из room.
 * Нужен механизм подключения обратно, если клиент "обновил" страницу или
 * переподключился к игре после disconnect(разрыв соединения)
 */

/**
 * Класс для объекта игры
 */
export default class GameService extends EventEmitter<{
  end: [
    {
      reason: string | undefined;
      statistic: Partial<Record<string, { exp: number; gold: number; nick: string }[]>>;
    },
  ];
  startOrders: [];
  endOrders: [];
  preKick: [{ reason: string; player: Player }];
  kick: [{ reason: string; player: Player }];
  startRound: [{ round: number; status: Record<string, GameStatus[]> }];
  endRound: [{ dead: Player[]; log: HistoryItem[] }];
}> {
  players: PlayersService;
  orders: OrderService;
  round = new RoundService();
  history = new HistoryService();
  longActions: Partial<Record<keyof typeof magics, LongItem[]>> = {};
  info!: Game;
  flags: {
    noDamageRound: number;
    global: GlobalFlags;
  };

  /**
   * Конструктор объекта игры
   * @param players массив игроков
   */
  constructor(players: string[]) {
    super();

    this.players = new PlayersService(players);
    this.orders = new OrderService(this.players, this.round);
    this.flags = {
      noDamageRound: 0,
      global: {
        isEclipsed: [],
      },
    };
  }

  /**
   * Функция проверки окончания игры
   */
  get isGameEnd(): boolean {
    return (
      this.isTeamWin ||
      this.players.alivePlayers.length === 0 ||
      this.flags.noDamageRound > 2 ||
      this.round.count > 9
    );
  }

  get isTeamWin(): boolean {
    const { [reservedClanName]: noClan, ...groupByClan } = this.players.groupByClan(
      this.players.alivePlayers,
    );
    if (!noClan?.length) {
      return Object.keys(groupByClan).length === 1;
    }
    return noClan.length === 1 && !Object.keys(groupByClan).length;
  }

  getEndGameReason() {
    if (this.flags.noDamageRound > 2) {
      return 'noDamageRound';
    }
  }

  get checkRoundDamage(): boolean {
    return !!this.history.hasDamageForRound(this.round.count);
  }

  /**
   * Предзагрузка игры
   */
  preLoading(): void {
    this.initHandlers();
    this.startGame();

    arena.games[this.info.id] = this;

    this.info.players.forEach((playerId) => {
      arena.characters[playerId].gameId = this.info.id;
    });
    // @todo add statistic +1 game for all players
  }

  /**
   * Старт игры
   */
  startGame(): void {
    console.debug('GC debug:: startGame', 'gameId:', this.info.id);
    this.round.initRound();
  }

  /**
   *@todo Остановка игры
   */
  pauseGame(): void {
    console.debug(this.info.id);
  }

  /**
   * Прекик, помечаем что пользователь не выполнил заказ и дальше будет выброшен
   * @param id id игрока, который будет помочен как бездействующий
   * @param reason строка, подставляющаяся в флаг isKicked
   */
  preKick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: preKick', id, 'no player');
      return;
    }
    player.preKick(reason);
    this.emit('preKick', { reason, player });
  }

  /**
   * Функция "выброса игрока" из игры без сохранения накопленных статов
   * @param id id игрока, который будет выброшен
   * @param reason причина кика
   */
  kick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: kick', id, 'no player');
      return;
    }

    this.emit('kick', { reason, player });

    const char = arena.characters[id];
    char.addGameStat({ runs: 1 });
    void char.saveToDb();
    char.autoreg = false;
    this.players.kick(id);
    this.info.players.splice(this.info.players.indexOf(id), 1);
    this.cleanLongMagics();
    char.gameId = '';
  }

  checkRun(player: Player) {
    if (player.flags.isKicked === 'run') {
      this.kick(player.id, player.flags.isKicked);
      return;
    }
  }

  /**
   * Проверяем делал ли игрок заказ. Помечает isKicked, если нет
   * @param player
   */
  checkOrders(player: Player): void {
    if (!player.alive || this.round.count === 1) {
      return;
    }

    if (!this.orders.checkPlayerOrderLastRound(player.id)) {
      if (player.flags.isKicked === 'afk') {
        this.kick(player.id, 'afk');
      } else {
        this.preKick(player.id, 'afk');
      }
    } else {
      /** @todo create clear kick flag method */
      player.preKick(undefined);
    }
  }

  /**
   * Ставим флаги, влияющие на окончание игры
   */
  handleEndGameFlags(): void {
    if (this.checkRoundDamage) {
      this.flags.noDamageRound = 0;
    } else {
      this.flags.noDamageRound += 1;
    }
  }

  recordOrderResult(item: HistoryItem) {
    this.history.addHistoryForRound(item, this.round.count);
  }

  getLastRoundResults() {
    return this.history.getHistoryForRound(this.round.count - 1);
  }

  getRoundResults() {
    return this.history.getHistoryForRound(this.round.count);
  }

  /**
   * @description Завершение игры
   *
   */
  endGame(): void {
    console.log('GC debug:: endGame', this.info.id);
    // Отправляем статистику
    setTimeout(async () => {
      const statistic = this.statistic();
      await this.saveGame();
      this.forAllPlayers((player: Player) => {
        arena.characters[player.id].gameId = '';
      });

      this.emit('end', { reason: this.getEndGameReason(), statistic });
      this.removeAllListeners();

      arena.mm.cancel();

      // FIXME move to client side
      this.forAllPlayers((player: Player) => {
        const char = arena.characters[player.id];
        if (char.expEarnedToday >= char.expLimitToday) {
          char.autoreg = false;
        }
        if (!char.autoreg) return;
        arena.mm.push({
          id: player.id,
          psr: 1000,
          startTime: Date.now(),
        });
      });
    }, 5000);
  }

  /**
   * Создание объекта в базе // потребуется для ведения истории
   * @return Объект созданный в базе
   */
  async createGame() {
    try {
      const dbGame = await createGame(this.players.init);
      this.info = dbGame;
      this.info.id = this.info._id.toString();
      return this;
    } catch (e) {
      console.log('GC debug:: createGame', e);
    } finally {
      this.preLoading();
    }
  }

  /**
   * Очищаем глобальные флаги в бою
   * затмение, бунт богов, и т.п
   */
  refreshRoundFlags(): void {
    this.flags.global.isEclipsed = [];
  }

  /**
   * Подвес
   */
  initHandlers(): void {
    // Обработка сообщений от Round Module
    this.round.subscribe(({ state, round }) => {
      switch (state) {
        case RoundStatus.START_ROUND: {
          this.forAllPlayers(this.checkOrders);
          this.sendStatus(round);
          break;
        }
        case RoundStatus.END_ROUND: {
          this.forAllPlayers(this.checkRun);
          this.emit('endRound', { dead: this.sortDead(), log: this.getRoundResults() });
          this.players.reset();
          this.orders.reset();
          this.handleEndGameFlags();
          if (this.isGameEnd) {
            this.round.unsubscribe();
            this.endGame();
          } else {
            this.refreshRoundFlags();
            this.round.nextRound();
          }
          break;
        }
        case RoundStatus.ENGINE: {
          engine(this);
          break;
        }
        case RoundStatus.START_ORDERS: {
          this.emit('startOrders');
          break;
        }
        case RoundStatus.END_ORDERS: {
          this.emit('endOrders');
          break;
        }
        default: {
          console.log('InitHandler:', state, 'undef event');
        }
      }
    });
  }

  /**
   * Метод сохраняющий накопленную статистику игроков в базу и сharObj
   * @todo нужен общий метод сохраняющий всю статистику
   */
  async saveGame() {
    try {
      const promises = this.info.players.map(async (id) => {
        const player = this.players.getById(id);
        if (!player) {
          return;
        }
        const isWinner = player.alive;

        await arena.characters[id].resources.addResources({
          gold: player.stats.collect.gold,
          exp: player.stats.collect.exp,
          components: player.stats.collect.component
            ? { [player.stats.collect.component]: 1 }
            : undefined,
        });

        const kills = this.players.getKills(id).length;
        const death = isWinner ? 0 : 1;

        arena.characters[id].addGameStat({
          games: 1,
          death,
          kills,
        });
        return arena.characters[id].saveToDb();
      });

      await Promise.all(promises);
    } catch (e) {
      console.log('Game:', e);
    }
  }

  /**
   * Функция послематчевой статистики
   * @todo надо как-то переделать функцию, потому что непонятно, что она мутирует данные
   * @return возвращает строку статистики по всем игрокам
   */
  statistic() {
    this.giveGoldforKill();
    const winners = this.players.alivePlayers;
    const gold = this.players.deadPlayers.length ? 5 : 1;
    winners.forEach((p) => {
      p.stats.addGold(gold);
      p.stats.addComponent(getRandomComponent(60));
    });

    this.players.deadPlayers.forEach((p) => {
      p.stats.addComponent(getRandomComponent(20));
    });

    const playersByClan = this.players.groupByClan();

    const getStatisticString = (p: Player) => ({
      nick: p.nick,
      exp: p.stats.collect.exp,
      gold: p.stats.collect.gold,
      component: p.stats.collect.component,
    });

    const statisticByClan = mapValues(playersByClan, (players) => {
      return players?.map(getStatisticString);
    });

    return statisticByClan;
  }

  /**
   * Функция пробегает всех убитых и раздает золото убийцам
   */
  giveGoldforKill(): void {
    this.players.deadPlayers.forEach((p) => {
      const killer = this.players.getById(p.getKiller());
      if (killer && killer.id !== p.id) {
        killer.stats.addGold(5 * p.lvl);
      }
    });
  }

  /**
   * Функция выставляет "смерть" для игроков имеющих hp < 0;
   * Отсылает сообщение о смерти игрока в последнем раунде
   * @todo сообщение о смерти как-то нормально нужно сделать,
   * чтобы выводило от чего и от кого умер игрок
   */
  sortDead() {
    const dead = this.players.sortDead();
    this.cleanLongMagics();
    return dead;
  }

  /**
   * Очистка массива длительных магий от умерших
   */
  cleanLongMagics(): void {
    this.longActions = mapValues(this.longActions, (longMagicType) => {
      return longMagicType?.filter((act) => {
        return this.players.getById(act.target)?.alive;
      });
    });
  }

  /**
   * Интерфейс для работы со всеми игроками в игре
   * @param f функция применяющая ко всем игрокам в игре
   */
  forAllPlayers(f: (player: Player) => void): void {
    this.players.players.forEach((p) => f.call(this, p));
  }

  getStatus() {
    const playersByClan = this.players.groupByClan();

    return mapValues(playersByClan, (players = []) => {
      return players.map((p) => p.getStatus());
    });
  }

  /**
   * Рассылка состояний живым игрокам
   */
  sendStatus(round: number): void {
    const status = this.getStatus();

    this.emit('startRound', { round, status });
  }
}
