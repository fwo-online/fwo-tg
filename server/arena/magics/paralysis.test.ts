import {
  describe, beforeAll, beforeEach, afterEach, it, spyOn, expect,
} from 'bun:test';
import casual from 'casual';
import { CharacterService } from '@/arena/CharacterService';
import GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import attack from '../actions/attack';
import paralysis from './paralysis';

// npm t server/arena/magics/paralysis.test.ts

describe('paralysis', () => {
  let game: GameService;

  beforeAll(() => {
    casual.seed(1);

    attack.registerPreAffects([paralysis]);
  });

  beforeEach(async () => {
    const initiator = await TestUtils.createCharacter({ prof: 'm', magics: { paralysis: 1 } });
    const target = await TestUtils.createCharacter({ prof: 'w' }, { withWeapon: true });

    await Promise.all([initiator.id, target.id].map(CharacterService.getCharacterById));

    game = new GameService([initiator.id, target.id]);
  });

  beforeEach(() => {
    spyOn(global.Math, 'random').mockReturnValue(0.1);
  });

  afterEach(() => {
    spyOn(global.Math, 'random').mockRestore();
  });

  it('initiator should be blinded by eclipse', async () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('base.mp', 99);
    game.players.players[0].stats.set('mp', 99);
    game.players.players[1].proc = 1;

    paralysis.cast(game.players.players[0], game.players.players[0], game);

    attack.cast(game.players.players[1], game.players.players[0], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
