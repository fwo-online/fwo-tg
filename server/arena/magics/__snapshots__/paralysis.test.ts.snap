// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paralysis initiator should be blinded by eclipse 1`] = `
"*asperiores* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _–ü–∞—Ä–∞–ª–∏—á_ –Ω–∞ *asperiores*
\\[ üìñ80 ]
*alias* –ø—ã—Ç–∞–ª—Å—è –∞—Ç–∞–∫–æ–≤–∞—Ç—å *asperiores*, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –ø–æ–ª—É—á–∏–ª–æ—Å—å
_–ü–∞—Ä–∞–ª–∏—á_ *asperiores*"
`;
